pallets/parachain-staking/src/lib.rs

```diff
@@ -834,34 +841,182 @@ pub mod pallet {
 		/// - Writes: CollatorState, CandidatePool, ExitQueue, N * AtStake,
 		///   SelectedCandidates
 		/// # </weight>
-		#[pallet::weight(<T as pallet::Config>::WeightInfo::leave_candidates(T::MaxCollatorCandidates::get(), T::MaxCollatorCandidates::get() * T::MaxDelegatorsPerCollator::get()))]
-		pub fn leave_candidates(origin: OriginFor<T>) -> DispatchResultWithPostInfo {
+		#[pallet::weight(<T as pallet::Config>::WeightInfo::init_leave_candidates(
+			T::MaxCollatorCandidates::get(),
+			T::MaxCollatorCandidates::get() * T::MaxDelegatorsPerCollator::get()
+		))]
+		pub fn init_leave_candidates(origin: OriginFor<T>) -> DispatchResultWithPostInfo {
 			let collator = ensure_signed(origin)?;
 			let mut state = <CollatorState<T>>::get(&collator).ok_or(Error::<T>::CandidateNotFound)?;
 			ensure!(!state.is_leaving(), Error::<T>::AlreadyLeaving);
-			let mut exits = <ExitQueue<T>>::get();
-			let now = <Round<T>>::get().current;
-			let when = now.saturating_add(T::ExitQueueDelay::get());
+			let mut candidates = <CandidatePool<T>>::get();
 			ensure!(
-				exits.insert(Stake {
-					owner: collator.clone(),
-					amount: when
-				}),
-				Error::<T>::AlreadyLeaving
+				candidates.len().saturating_sub(1) as u32 >= T::MinSelectedCandidates::get(),
+				Error::<T>::TooFewCollatorCandidates
 			);
+
+			let now = <Round<T>>::get().current;
+			let when = now.saturating_add(T::ExitQueueDelay::get());
 			state.leave_candidates(when);
-			let mut candidates = <CandidatePool<T>>::get();
 			if candidates.remove_by(|stake| stake.owner.cmp(&collator)).is_some() {
 				<CandidatePool<T>>::put(candidates);
 			}
-			<ExitQueue<T>>::put(exits);
 			<CollatorState<T>>::insert(&collator, state);
 
 			// update candidates for next round
 			let (num_collators, num_delegators, _, _) = Self::select_top_candidates();
 
 			Self::deposit_event(Event::CollatorScheduledExit(now, collator, when));
-			Ok(Some(<T as pallet::Config>::WeightInfo::leave_candidates(
+			Ok(Some(<T as pallet::Config>::WeightInfo::init_leave_candidates(
```

Comment:
Especially if we rare returning it here anyway, no?

Reply:
Ok I guess it depends on `select_top_candidates` being executed, alright. 😄

Reply:
> Ok I guess it depends on `select_top_candidates` being executed, alright. 😄

exactly 💯